### PR TITLE
Element.checkVisibility() - docs

### PR DESCRIPTION
**Pull request by hamishwillee**
FF122 adds support to get whether an element's rendering is been skipped due to `content-visibility:auto` being set. The support is a new property of the `Element.checkVisibility()` options parameter object named `contentVisibilityAuto`. There are also two other options   `opacityProperty` and `visibilityProperty` created as aliases of other existing options in the same bugzilla https://bugzilla.mozilla.org/show_bug.cgi?id=1859852

This adds docs for the `checkVisibility()` method.

Other docs work for this can be tracked in #31107

Note, the example seems to show some issue with the default behaviour for checkVisibility on Firefox. See https://bugzilla.mozilla.org/show_bug.cgi?id=1859852#c7

### CODE REVIEW

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+      - : `true` to check if the element is invisible due to its CSS [`visibility`](/en-US/docs/Web/CSS/visibility) value.
+        `false` by default.
+
+        > **Note:** Invisible elements include those that have [`visibility: hidden`](/en-US/docs/Web/CSS/visibility#hidden), and some element types that have [`visibility: collapse`](/en-US/docs/Web/CSS/visibility#collapse).
+
+    - `checkOpacity`
+      - : A historic alias for [`opacityProperty`](#opacityProperty).
+    - `checkVisibilityCSS`
+      - : A historic alias for [`visibilityProperty`](#visibilityProperty).
```
Comment (teoli2003) [inline files/en-us/web/api/element/checkvisibility/index.md:48]: Drive-by comment: internal anchors must be lowercase.
```suggestion
      - : A historic alias for [`opacityProperty`](#opacityproperty).
    - `checkVisibilityCSS`
      - : A historic alias for [`visibilityProperty`](#visibilityproperty).
```

Comment (hamishwillee) [reply files/en-us/web/api/element/checkvisibility/index.md:48]: Thanks. Must be dim - couldn't work out why this was showing a flaw.

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+page-type: web-api-instance-method
+browser-compat: api.Element.checkVisibility
+---
+
+{{APIRef("DOM")}}
+
+The **`checkVisibility()`** method of the {{domxref("Element")}} interface performs checks to determine whether the element is visible.
+
+The method returns `false` if the element doesn't have an associated box, such as those that have the CSS property [`display: none`](/en-US/docs/Web/CSS/display#none) or [`display: contents`](/en-US/docs/Web/CSS/display#contents).
+The method also returns `false` if the element is not being rendered because the element or an an ancestor element sets the property value as [`content-visibility: hidden`](/en-US/docs/Web/CSS/content-visibility#hidden).
```
Comment (wbamberg) [inline files/en-us/web/api/element/checkvisibility/index.md:14]: Strictly optional, but I favour bullets for these situations. Also, the wording "the CSS property [display: none](https://pr31322.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/display#none) or [display: contents](https://pr31322.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/display#contents)" (and similar in other parts of this page) is a bit weird here. `display` is the property, `none` is a value. So I think "the `display` property set to `none` or `contents`" seems more natural to me, and similarly for the others.

For example:

```suggestion
The method returns `false` in either of the following situations:
- The element doesn't have an associated box, for example because the CSS {{cssxref("display")}} property is set to [`none`](/en-US/docs/Web/CSS/display#none) or [`contents`](/en-US/docs/Web/CSS/display#contents).
- The element is not being rendered because the element or an an ancestor element sets the {{cssxref("content-visibility")}} property to [`hidden`](/en-US/docs/Web/CSS/content-visibility#hidden).
```

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+The optional parameter enables additional checks to test for other interpretations of what "visible" means.
+For example, you can further check whether an element has an opacity of `0`, if the CSS [`visibility`](/en-US/docs/Web/CSS/visibility) means that the element is invisible, or if its rendering is currently being skipped because it has the property [`content-visibility: auto`](/en-US/docs/Web/CSS/content-visibility#auto).
```
Comment (wbamberg) [inline files/en-us/web/api/element/checkvisibility/index.md:17]:
```suggestion
For example, you can further check whether an element has an opacity of `0`, if the CSS [`visibility`](/en-US/docs/Web/CSS/visibility) means that the element is invisible, or if its rendering is currently being skipped because it has the {{cssxref("content-visibility")}} property set to [`auto`](/en-US/docs/Web/CSS/content-visibility#auto).
```

etc etc.

Also though am I correct in understanding that "if its rendering is currently being skipped because it has the `content-visibility` property set to `auto`" is not quite correct because skipping rendering also requires that the element is "not relevant"?

So would just "if its rendering is currently being skipped because it has the `content-visibility` property set to `auto`" be more accurate?

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+### Parameters
+
+- `options` {{optional_inline}}
+
+  - : An object indicating additional checks to run.
+    The possible options are:
+
+    - `contentVisibilityAuto`
+      - : `true` to check if the element has [`content-visibility: auto`](/en-US/docs/Web/CSS/content-visibility#auto) and is currently skipping its rendering.
```
Comment (wbamberg) [inline files/en-us/web/api/element/checkvisibility/index.md:33]: Oh this implies that you need both, which makes the "because" above misleading?

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    - `checkOpacity`
+      - : A historic alias for [`opacityProperty`](#opacityproperty).
+    - `checkVisibilityCSS`
+      - : A historic alias for [`visibilityProperty`](#visibilityproperty).
+
+### Return value
+
+`false` if any of the following conditions are met, otherwise `true`:
+
+- the element doesn't have an associated box
```
Comment (wbamberg) [inline files/en-us/web/api/element/checkvisibility/index.md:54]:
```suggestion
- The element doesn't have an associated box.
```

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+- The element has [`content-visibility: hidden`](/en-US/docs/Web/CSS/visibility#hidden).
+- `opacityProperty` (or `checkOpacity`) is `true` and the element has an opacity of `0`.
+- `visibilityProperty` (or `checkVisibilityCSS`) is `true` and the element is invisible due to the CSS [`visibility`](/en-US/docs/Web/CSS/visibility) property.
+- `contentVisibilityAuto` is `true` and element rendering is being skipped due to [`content-visibility: auto`](/en-US/docs/Web/CSS/content-visibility#auto)
```
Comment (wbamberg) [inline files/en-us/web/api/element/checkvisibility/index.md:58]:
```suggestion
- `contentVisibilityAuto` is `true` and element rendering is being skipped due to [`content-visibility: auto`](/en-US/docs/Web/CSS/content-visibility#auto).
```

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+```
+
+#### Result
+
+The results are shown below.
+If you change the selection the results will be applied to the test element (blue outline) and the results of the `checkVisibility()` for each setting should be displayed.
+So for example, if you set the `opacity: 0` that test (only) should indicate `false`.
+However if you set `display: none` then all tests should return `false.
+
+{{ EmbedLiveSample('Using hidden to manually manage visibility', "100%", "200" ) }}
```
Comment (wbamberg) [inline files/en-us/web/api/element/checkvisibility/index.md:177]: The argument here is wrong (it works because EmbedLiveSample is very lax).

Comment (wbamberg) [inline files/en-us/web/api/element/checkvisibility/index.md:177]: This is a nice example. Just a suggestion, but it might be a bit easier to follow if you actually list the API call signature, instead of:


> * default: true
> * opacity: true
> * visibility: true
> * contentVisibilityAuto: true

like:


> * Result of `checkVisibility()`: true
> * Result of `checkVisibility({opacityProperty: true})`: true
> * Result of `checkVisibility({visibilityProperty: true})`: true
> * Result of `checkVisibility({contentVisibilityAuto: true})`: true

**Review by wbamberg — COMMENTED** [review]
I had some copy-edit type suggestions.

The only substantive thing I thought was: AIUI if we have `content-visibility: auto`, then the element is not rendered only if it is not relevant. If we pass `contentVisibilityAuto: true`, and the element has `content-visibility: auto` set, but it is being rendered, is that going to return true or false?

File: files/en-us/web/api/element/checkvisibility/index.md (hunk 2 above)
Comment (hamishwillee) [reply files/en-us/web/api/element/checkvisibility/index.md:14]: Thanks - agreed.

File: files/en-us/web/api/element/checkvisibility/index.md (hunk 3 above)
Comment (hamishwillee) [reply files/en-us/web/api/element/checkvisibility/index.md:17]: The content is skipped if `content-visibility: auto` and the specific content is outside the viewport. 

```suggestion
For example, you can further check whether an element has an opacity of `0`, if the value of the {{cssxref("visibility")}} property means that the element is invisible, or if its rendering is currently being skipped because it has the {{cssxref("content-visibility")}} property set to [`auto`](/en-US/docs/Web/CSS/content-visibility#auto) and it is not in the visible viewport).
```

File: files/en-us/web/api/element/checkvisibility/index.md (hunk 4 above)
Comment (hamishwillee) [reply files/en-us/web/api/element/checkvisibility/index.md:33]: Yes, you need both. This is correct - I adjusted the comment above.

File: files/en-us/web/api/element/checkvisibility/index.md (hunk 6 above)
Comment (hamishwillee) [reply files/en-us/web/api/element/checkvisibility/index.md:58]: Note that when I say "property is", set on the element or an ancestor (i.e. the computed value). This is implied right?

```suggestion
- `contentVisibilityAuto` is `true`, the {{cssxref("content-visibility")}} property is [`auto`](/en-US/docs/Web/CSS/content-visibility#auto), and element rendering is being skipped.
```

File: files/en-us/web/api/element/checkvisibility/index.md (hunk 4 above)
Comment (hamishwillee) [reply files/en-us/web/api/element/checkvisibility/index.md:33]:
```suggestion
      - : `true` to check if the element's {{cssxref("content-visibility")}} property has (or inherits) the value [`auto`](/en-US/docs/Web/CSS/content-visibility#auto) and it is currently skipping its rendering.
```

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+- `options` {{optional_inline}}
+
+  - : An object indicating additional checks to run.
+    The possible options are:
+
+    - `contentVisibilityAuto`
+      - : `true` to check if the element has [`content-visibility: auto`](/en-US/docs/Web/CSS/content-visibility#auto) and is currently skipping its rendering.
+        `false` by default.
+    - `opacityProperty`
+      - : `true` to check if the element has an opacity of `0`.
```
Comment (hamishwillee) [inline files/en-us/web/api/element/checkvisibility/index.md:36]:
```suggestion
      - : `true` to check if the element {{cssxref("opacity")}} property is set to (or inherits) a value of `0`.
```

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        `false` by default.
+    - `visibilityProperty`
+
+      - : `true` to check if the element is invisible due to its CSS [`visibility`](/en-US/docs/Web/CSS/visibility) value.
```
Comment (hamishwillee) [inline files/en-us/web/api/element/checkvisibility/index.md:40]:
```suggestion
      - : `true` to check if the element is invisible due to the value of its {{cssxref("visibility")}} property.
```

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    - `checkVisibilityCSS`
+      - : A historic alias for [`visibilityProperty`](#visibilityproperty).
+
+### Return value
+
+`false` if any of the following conditions are met, otherwise `true`:
+
+- the element doesn't have an associated box
+- The element has [`content-visibility: hidden`](/en-US/docs/Web/CSS/visibility#hidden).
+- `opacityProperty` (or `checkOpacity`) is `true` and the element has an opacity of `0`.
```
Comment (hamishwillee) [inline files/en-us/web/api/element/checkvisibility/index.md:56]:
```suggestion
- `opacityProperty` (or `checkOpacity`) is `true` and the element {{cssxref("opacity ")}} property has (or inherits) a value of `0`.
```

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+- `visibilityProperty` (or `checkVisibilityCSS`) is `true` and the element is invisible due to the CSS [`visibility`](/en-US/docs/Web/CSS/visibility) property.
```
Comment (hamishwillee) [inline files/en-us/web/api/element/checkvisibility/index.md:57]:
```suggestion
- `visibilityProperty` (or `checkVisibilityCSS`) is `true` and the element is invisible due to the value of its {{cssxref("visibility")}} property.
```

File: files/en-us/web/api/element/checkvisibility/index.md (hunk 7 above)
Comment (hamishwillee) [reply files/en-us/web/api/element/checkvisibility/index.md:177]:
```suggestion
{{ EmbedLiveSample('Test checkVisibility() with varied CSS', "100%", "200" ) }}
```

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+page-type: web-api-instance-method
+browser-compat: api.Element.checkVisibility
+---
+
+{{APIRef("DOM")}}
+
+The **`checkVisibility()`** method of the {{domxref("Element")}} interface checks whether the element is visible.
+
+The method returns `false` in either of the following situations:
+- The element doesn't have an associated box, for example because the CSS {{cssxref("display")}} property is set to [`none`](/en-US/docs/Web/CSS/display#none) or [`contents`](/en-US/docs/Web/CSS/display#contents).
```
[the remark this answers is not in the document]
Comment (hamishwillee) [reply files/en-us/web/api/element/checkvisibility/index.md:15]: This comment seems invalid

**Comment by hamishwillee** [comment]
Thanks very much @wbamberg - I have accepted your comments in either letter or spirit and rolled out the general comments consistently. Ready for another look.

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+```js
+const displayCssSelect = document.getElementById("css_display");
+const contentVisibilityCssSelect = document.getElementById(
+  "css_content_visibility",
+);
+const displayCssOpacity = document.getElementById("css_opacity");
+const displayCssVisibility = document.getElementById("css_visibility");
+
+const visibility_check = document.getElementById("visibility_check");
```
Comment (wbamberg) [inline files/en-us/web/api/element/checkvisibility/index.md:130]: Sorry I missed this before but snake case is against our coding guidelines, so this ought to be `visibilityCheck`.

File: files/en-us/web/api/element/checkvisibility/index.md [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+  });
+  const cssVisibilityCheck = elementToCheck.checkVisibility({
+    visibilityProperty: true,
+  });
+  const contentVisibilityAutoCheck = elementToCheck.checkVisibility({
+    contentVisibilityAuto: true,
+  });
+
+  // Output the results of the tests
+  visibility_check.innerHTML = `<p>Checks on element below (may be hidden):</p>
```
Comment (wbamberg) [inline files/en-us/web/api/element/checkvisibility/index.md:162]: I also missed this before and I can see why you did it but I would not have `innerHTML` here. I think it is too often an unsafe API for us to include it in our examples. See also https://github.com/mdn/content/issues/13496.

In situations like this I just use a `<pre>` block and set text content, it's not as pretty but shorter and safer.

But like I say I can see why you did it so won't insist on it, just registering my dissatisfaction with it :).

**Review by wbamberg — COMMENTED** [review]
Thank you for the updates Hamish!

I had a couple of comments - the variable naming one should be fixed as [it's clear in our coding guidelines](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Writing_style_guide/Code_style_guide/JavaScript#object_names).

The other one, about innerHTML, I don't like it but would not insist on it.

File: files/en-us/web/api/element/checkvisibility/index.md (hunk 3 above)
Comment (wbamberg) [reply files/en-us/web/api/element/checkvisibility/index.md]: It's much clearer now 👍 .

File: files/en-us/web/api/element/checkvisibility/index.md (hunk 13 above)
Comment (hamishwillee) [reply files/en-us/web/api/element/checkvisibility/index.md]: Absolutely. Thanks.

File: files/en-us/web/api/element/checkvisibility/index.md (hunk 14 above)
Comment (hamishwillee) [reply files/en-us/web/api/element/checkvisibility/index.md]: Thanks @wbamberg 

I've modified this to a `<pre>` too. It is shorter and safer. I want to argue that we should work to show developers the "right way" to safely set/sanitize content rather than adjust our examples to always do the safe way, but that's not the most pragmatic approach. Too easy to get wrong, and not really necessary when the point of the example isn't demonstrating "UI stuff".

I wonder if perhaps it is time to update our sections on code example, and in particular this one for live examples: https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Live_samples#grouping_code_blocks_by_heading

We could have a live example which uses innerText and a `<pre>` block, along with a log function or similar, explain why, and hopefully authors will copy this rather than some other example.

Comment (hamishwillee) [reply files/en-us/web/api/element/checkvisibility/index.md]: PS Thanks for taking the time to go through this properly. I think I've already mentally checked out.

Comment (wbamberg) [reply files/en-us/web/api/element/checkvisibility/index.md]: > Thanks @wbamberg
> 
> I've modified this to a `\<pre\>` too. It is shorter and safer. I want to argue that we should work to show developers the "right way" to safely set/sanitize content rather than adjust our examples to always do the safe way, but that's not the most pragmatic approach. Too easy to get wrong, and not really necessary when the point of the example isn't demonstrating "UI stuff".

That's an interesting suggestion, but I do think it's best for examples to minimise nonessential complexity. People are here to learn about `checkVisibility()` and we should concentrate on that. IMO.

> I wonder if perhaps it is time to update our sections on code example, and in particular this one for live examples: https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Live_samples#grouping_code_blocks_by_heading
> 
> We could have a live example which uses innerText and a `\<pre\>` block, along with a log function or similar, explain why, and hopefully authors will copy this rather than some other example.

It's a good idea. Better would be for Yari to support a console for live samples, (see https://github.com/orgs/mdn/discussions/215) but, you know...

**Review by wbamberg — APPROVED** [review]
👍 thank you @hamishwillee , looks great!

File: files/en-us/web/api/element/checkvisibility/index.md (hunk 14 above)
Comment (hamishwillee) [reply files/en-us/web/api/element/checkvisibility/index.md]: Thanks for the review.

Yeah, having infrastructure support might be cool. Except that the infrastructure we currently have for these makes them practically un-debuggable, and I'm scared adding "more" would break this "more".
So a docs fix: https://github.com/mdn/content/pull/31596
